lbrynet/core/server/DHTHashAnnouncer.py

```diff
@@ -45,60 +73,102 @@ def immediate_announce(self, blob_hashes):
+            if not self._total:
+                self._total = len(hashes)
 
-        for h in hashes:
-            announce_deferred = defer.Deferred()
-            ds.append(announce_deferred)
-            if immediate:
-                self.hash_queue.appendleft((h, announce_deferred))
-            else:
-                self.hash_queue.append((h, announce_deferred))
         log.debug('There are now %s hashes remaining to be announced', self.hash_queue_size())
 
-        def announce():
+        @defer.inlineCallbacks
+        def do_store(blob_hash, announce_d):
+            if announce_d.called:
+                defer.returnValue(announce_deferred.result)
+            try:
+                store_nodes = yield self.dht_node.announceHaveBlob(binascii.unhexlify(blob_hash),
+                                                                   self.peer_port)
+                if not store_nodes:
+                    retries = self._retries.get(blob_hash, 0)
+                    retries += 1
+                    self._retries[blob_hash] = retries
+                    if retries <= self.STORE_RETRIES:
+                        log.debug("No nodes stored %s, retrying", blob_hash)
+                        result = yield do_store(blob_hash, announce_d)
+                    else:
+                        log.warning("No nodes stored %s", blob_hash)
+                else:
+                    result = store_nodes
+                if not announce_d.called:
+                    announce_d.callback(result)
+                defer.returnValue(result)
+            except Exception as err:
+                if not announce_d.called:
+                    announce_d.errback(err)
+                raise err
+
+        @defer.inlineCallbacks
+        def announce(progress=None):
+            progress = progress or {}
             if len(self.hash_queue):
-                h, announce_deferred = self.hash_queue.popleft()
-                log.debug('Announcing blob %s to dht', h)
-                d = self.dht_node.announceHaveBlob(binascii.unhexlify(h), self.peer_port)
-                d.chainDeferred(announce_deferred)
-                d.addBoth(lambda _: utils.call_later(0, announce))
+                with self._lock:
+                    h, announce_deferred = self.hash_queue.popleft()
+                log.debug('Announcing blob %s to dht', h[:16])
+                stored_to_nodes = yield do_store(h, announce_deferred)
+                progress[h] = stored_to_nodes
+                log.debug("Stored %s to %i peers (hashes announced by this announcer: %i)",
+                          h.encode('hex')[:16],
+                          len(stored_to_nodes), len(progress))
+
+                yield announce(progress)
             else:
-                self._concurrent_announcers -= 1
+                with self._lock:
+                    self._concurrent_announcers -= 1
+            defer.returnValue(progress)
 
```

Comment:
We can use a defer.DeferredSemaphore to simplify this and remove managing of self._concurrent_announcers 

See https://github.com/lbryio/lbry/blob/bbc960aaf95569d4cafcbae7e4b1c638973f3797/lbrynet/file_manager/EncryptedFileManager.py#L245 for example

Reply:
Actually I think I need to think about this a bit, there might be a cleaner way of doing this using DeferredQueue's